src/pip/_internal/commands/download.py

```diff
@@ -134,7 +140,7 @@ def run(self, options, args):
 
         downloaded = ' '.join([
             req.name for req in requirement_set.requirements.values()
-            if req.successfully_downloaded
+            if req.successfully_downloaded and req.name
```

Comment:
It was also to avoid a surprise since we are joining the list of `req.name` and if it is None,  we will get a `TypeError: sequence item 1: expected str instance, NoneType found` which can be avoided by doing this.

Still if this makes you feel uncomfortable that this can affect something else in the flow, I am happy to add a `# type: ignore`

Reply:
I think we *want* to have a surprise if we reach this point with a null `req.name`.

Reply:
Got it, but I am confused about where to put `# type: ignore` in this multi-line statement.

I tried a few different combinations, but only

```
        downloaded = ' '.join([ # type: ignore
            req.name for req in requirement_set.requirements.values() # type: ignore
            if req.successfully_downloaded
        ])
```

seems to make both mypy Python 2 and Python 3 happy.

Reply:
For now I have formatted it to put `type: ignore` in one line in https://github.com/pypa/pip/pull/8265/commits/5e33373a0738a03ca8515e3618046f0ad4b10582